### PR TITLE
System.Windows.Forms.DataGridView.RefreshColumns() improvements

### PR DESCRIPTION
**Pull request by kirsan31**
## Proposed changes
- Not generate bound columns if `AutoGenerateColumns == false`.
- Removed double memory allocation in `GetCollectionOfBoundDataGridViewColumns`.

## Customer Impact

- Performance improvements in `AutoGenerateColumns == false`  - no need to generate bound columns in this mode.
- Performance improvements in `AutoGenerateColumns == true`  - no more double memory allocation and coping.

## Regression? 

- No

## Risk

- Minimal.

## Test methodology

Existing tests and manual testing.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6956)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.DataConnection.cs [hunk 1]
```diff
@@ -991,14 +991,14 @@ private static DataGridViewColumn GetDataGridViewColumnFromType(Type type)
...
-            public DataGridViewColumn[] GetCollectionOfBoundDataGridViewColumns()
+            public List<DataGridViewColumn> GetCollectionOfBoundDataGridViewColumns()
             {
                 if (_props is null)
                 {
                     return null;
                 }
 
-                ArrayList cols = new ArrayList();
+                List<DataGridViewColumn> cols = new List<DataGridViewColumn>();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.DataConnection.cs:1001]: nit
```suggestion
                List<DataGridViewColumn> columns = new();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs [hunk 2]
```diff
@@ -1624,7 +1624,7 @@ private void AutoGenerateDataBoundColumns(DataGridViewColumn[] boundColumns)
             // 4. Add new columns for the Fields which were not data bound previously ( ie, for fields which do not have a clone ).
             if (boundColumns is not null)
             {
-                for (int j = 0; j < boundColumns.Length; j++)
+                for (int j = 0; j < boundColumns.Count; j++)
                 {
                     if (boundColumns[j] is not null && boundColumns[j].IsBrowsableInternal)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:1629]: Slight OT.
I think we can further simplify this, though NRT will make it obvious.
We know that `AutoGenerateDataBoundColumns` has a single callsite - `RefreshColumns`. The input parameter can either be `null`, or comes from `DataConnection.GetCollectionOfBoundDataGridViewColumns()` call - which can either be `null` or a list of non-nulls. This means `boundColumns[j] is not null` is always `true`.

**Review by RussKie — APPROVED** [review]

**Review by RussKie — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.DataConnection.cs [hunk 3]
```diff
@@ -1032,9 +1032,7 @@ public DataGridViewColumn[] GetCollectionOfBoundDataGridViewColumns()
                     cols.Add(dataGridViewColumn);
                 }
 
-                DataGridViewColumn[] ret = new DataGridViewColumn[cols.Count];
-                cols.CopyTo(ret);
-                return ret;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.DataConnection.cs:1037]: by any chance, this copy was modifiable earlier and now not without impacting the original columns?

Comment (kirsan31) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.DataConnection.cs:1037]: We have not here _original_ columns. We create here new List and fill it with new created columns. 
But in any case, it is then not modified.

**Comment by coderb** [comment]
Please use the list constructor that takes a capacity when it is known.  This can save memory and also avoid garbage.
`List<DataGridViewColumn> cols = new();`  => `List<DataGridViewColumn> cols = new(_props.Count);`

**Comment by kirsan31** [comment]
> Please use the list constructor that takes a capacity when it is known. This can save memory and also avoid garbage. `List<DataGridViewColumn> cols = new();` => `List<DataGridViewColumn> cols = new(_props.Count);`

I thought about it too, but capacity will not be `_props.Count` in all cases (may be less) and I decided to leave original realization...

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs [hunk 4]
```diff
@@ -1624,39 +1624,41 @@ private void AutoGenerateDataBoundColumns(DataGridViewColumn[] boundColumns)
...
-                        if (addNewColumn)
-                        {
-                            dataGridViewCols.Add(boundColumns[j]);
-                        }
-                        else
+                    bool addNewColumn = true;
+                    // Go thru the list of cloned columns and see if there is another column w/ the same data property name.
+                    int clonedColIndex = 0;
+                    for (; clonedColIndex < clonedColumnsCount; clonedColIndex++)
+                    {
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:1638]: can you  simplify this block by moving  `finalClonedColumns[clonedColIndex]` in to local ? also, you won't need `clonedIndex` outside of the for loop then.

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:1638]: i did not see local for `finalClonedColumns[clonedColIndex]`.  You are accessing indexed location multiple times. why not store it local and reuse?

Comment (kirsan31) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.Methods.cs:1638]: So I misunderstood you the first time...

**Review by RussKie — APPROVED** [review]

**Pull request merged**